### PR TITLE
Replacing custom encryption with SSL

### PR DESCRIPTION
**Pull request by jim-sheldon**

### CODE REVIEW

File: e2e/setup_localstack.py [hunk 1]
```diff
@@ -110,20 +118,26 @@
...
 		logging.info("Created secret")
 
 
+def setup_acm():
+	logging.info("Creating AWS ACM certificates")
+	for domain_name in [CERT_DOMAIN_NAME_A, CERT_DOMAIN_NAME_B, CERT_DOMAIN_NAME_C]:
+		logging.info(f"Created certificate for domain {domain_name}")
+		response = ACM_CLIENT.request_certificate(DomainName=domain_name)
+		certificate_arn = response.get("CertificateArn")
+		logging.info(f"Created certificate with ARN {certificate_arn}")
```
[the remark this answers is not in the document]
Comment (SpencerMarks-IDHA) [reply e2e/setup_localstack.py:134]: this looks like a valid concern.

File: gh/amqp_server.py [hunk 2]
```diff
@@ -84,15 +72,13 @@ def add_curation_data(partner_name: str, curation_data: dict, auto_approve: bool
...
 
 
-def run_get_cases_job(pathogen_config: PathogenConfig, partner: Partner, metadata: list[tuple], key: bytes):
+def run_get_cases_job(pathogen_config: PathogenConfig, partner: Partner, metadata: list[tuple]):
 	logging.info(f"Getting cases for pathogen {pathogen_config.name}")
 	proto_cases = get_partner_cases(pathogen_config.name, partner, metadata)
 	dict_cases = MessageToDict(proto_cases, preserving_proto_field_name=True).get("cases")
-	logging.debug(f"Encrypted new cases: {dict_cases}")
+	logging.debug(f"New cases: {dict_cases}")
 	# TODO: If no data log critical and skip
```
Comment (SpencerMarks-IDHA) [inline gh/amqp_server.py:80]: we try 
1) not to have to do comments in code that is going into production 
2) when an exception occurs or a critical error happens, typically just logging (and easily forgetting about it) is a bad idea.

File: gh/aws.py [hunk 3]
```diff
@@ -54,6 +41,32 @@
...
+	acm_client = boto3.client("acm", region_name=AWS_REGION)
+	if LOCALSTACK_URL:
+		acm_client = boto3.client("acm", endpoint_url=LOCALSTACK_URL, region_name=AWS_REGION)
+	response = acm_client.list_certificates()
+	logging.debug(f"Response: {response}")
+	certificates_list = response.get("CertificateSummaryList", [])
+
+	certificate_arn = ""
+	for certificate in certificates_list:
+		logging.debug(f"Certificate: {certificate}")
```
[the remark this answers is not in the document]
Comment (SpencerMarks-IDHA) [reply gh/aws.py:55]: please check

File: gh/fake_grpc_service.py [hunk 4]
```diff
@@ -22,17 +21,56 @@
...
+
+def get_certificate_arn() -> str:
+    logging.info("Getting certs from AWS")
+    response = ACM_CLIENT.list_certificates()
+    logging.info(f"Response: {response}")
+    certificates_list = response.get("CertificateSummaryList", [])
+    for certificate in certificates_list:
+        logging.info(f"Certificate: {certificate}")
+        if certificate.get("DomainName") == CERT_DOMAIN_NAME:
+            return certificate.get("CertificateArn", "")
```
Comment (SpencerMarks-IDHA) [inline gh/fake_grpc_service.py:71]: returning an empty string is problematic.

**Review by SpencerMarks-IDHA — COMMENTED** [review]
please address comments, thanks. also adding Victor to additional eyes on this code. thanks.

**Comment by SpencerMarks-IDHA** [comment]
Hi, Please give me admin perms on this github account and also please add [victormoyanodiaz] so he can help with code reviews. thanks!

**Comment by SpencerMarks-IDHA** [comment]
also, please update deps.

File: gh/fake_grpc_service.py [hunk 5]
```diff
@@ -22,17 +21,56 @@
...
+
+CERT_DOMAIN_NAME = os.environ.get("ACM_CERT_DOMAIN_NAME", "fake_grpc_server")
+DECRYPTION_PASSPHRASE = os.environ.get("DECRYPTION_PASSPHRASE", "foobar")
 
 JWKS_HOST = os.environ.get("JWKS_HOST")
 JWKS_FILE = os.environ.get("JWKS_FILE")
 
 
+def get_server_credentials():
+    logging.info("Getting server credentials")
```
[the remark this answers is not in the document]
Comment (jim-sheldon) [reply gh/fake_grpc_service.py:40]: Sure, can you grant access and I will review?

File: gh/amqp_server.py (hunk 2 above)
Comment (jim-sheldon) [reply gh/amqp_server.py]: I like both of those. I updated this block to log `warning` and return. `critical` may be too high.

File: gh/fake_grpc_service.py (hunk 4 above)
Comment (jim-sheldon) [reply gh/fake_grpc_service.py]: Yeah, I think it's better to `raise` in the event of no cert.

File: gh/fake_grpc_service.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jim-sheldon) [reply gh/fake_grpc_service.py:40]: Opening another PR for docstrings, will rebase when merged.

[the remark this answers is not in the document]
Comment (jim-sheldon) [reply gh/fake_grpc_service.py:42]: I added type hints and docstrings.

**Comment by jim-sheldon** [comment]
> also, please update deps.

All set

**Comment by jim-sheldon** [comment]
> please address comments, thanks. also adding Victor to additional eyes on this code. thanks.

Thanks for reviewing and providing feedback. I invited him to the repo, and will add him as a reviewer once the invitation gets accepted.

File: e2e/setup_localstack.py [hunk 6]
```diff
@@ -155,9 +159,23 @@
...
+    """
+    Setup AWS ACM resources
+    """
+
+    logging.info("Creating AWS ACM certificates")
+    for domain_name in [CERT_DOMAIN_NAME_A, CERT_DOMAIN_NAME_B, CERT_DOMAIN_NAME_C]:
+        logging.info(f"Created certificate for domain {domain_name}")
+        response = ACM_CLIENT.request_certificate(DomainName=domain_name)
+        certificate_arn = response.get("CertificateArn")
+        logging.info(f"Created certificate with ARN {certificate_arn}")
```
[the remark this answers is not in the document]
Comment (victormoyanodiaz) [reply e2e/setup_localstack.py:172]: I agree we should not be logging anything related with the certificates. Even though you are not logging the certificate itself here, we should probably not be logging anything related to it

**Review by victormoyanodiaz — CHANGES_REQUESTED** [review]
Hey Jim! It's really nice to meet you.

It seems like in this PR you are trying to replace the encryption mechanism from a custom encryption mechanism to SSL, but while doing that you also did some additional changes that have nothing to do with that particular goal. 

When working in something specific (and delicate) as changing the encryption mechanism, I'd say you want to keep your PRs as small as possible, with the minimum changes possible. You want to do it like that because if something does not work as expected, it is way easier to find out what's wrong with a PR that only has 8 or 9 files changed, than another one that has 30.

And please, believe me when I say that this thing happens to me every time. I always end up with monster PRs too. But when doing a change like the one you are trying to do, we need to try to keep it small or else it's harder debug issues later on.

On the other hand, I have not run the code and can not tell if it works or not, but the code in general looks good. Happy to approve the pr if we remove the certificate logs

File: e2e/setup_localstack.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jim-sheldon) [reply e2e/setup_localstack.py:172]: This script sets up mock AWS resources for local development and testing only.

**Comment by jim-sheldon** [comment]
> Hey Jim! It's really nice to meet you.
> 
> It seems like in this PR you are trying to replace the encryption mechanism from a custom encryption mechanism to SSL, but while doing that you also did some additional changes that have nothing to do with that particular goal.
> 
> When working in something specific (and delicate) as changing the encryption mechanism, I'd say you want to keep your PRs as small as possible, with the minimum changes possible. You want to do it like that because if something does not work as expected, it is way easier to find out what's wrong with a PR that only has 8 or 9 files changed, than another one that has 30.
> 
> And please, believe me when I say that this thing happens to me every time. I always end up with monster PRs too. But when doing a change like the one you are trying to do, we need to try to keep it small or else it's harder debug issues later on.
> 
> On the other hand, I have not run the code and can not tell if it works or not, but the code in general looks good. Happy to approve the pr if we remove the certificate logs

Hi, nice to meet you too, and thank you for reviewing my PR.

I agree, and apologize this is not a small PR. I too prefer incremental changes and will try to keep things that way. It makes things easier to write and read.

I also want things easily-debuggable, which is why I wrote the code using TDD, focused on integration and end-to-end  tests. In practice, this should let you use Github Actions as a free user in perpetuity, and if I should add tests somewhere, definitely flag it, and I will write and automate them. Local runs work too, you just need a Localstack API key for them, and populated `.env` files.

**Review by victormoyanodiaz — APPROVED** [review]
LGTM

**Review by SpencerMarks-IDHA — CHANGES_REQUESTED** [review]
Just don't print the certs and we are good here.

**Comment by jim-sheldon** [comment]
> Just don't print the certs and we are good here.

What remains is diagnostics for mocks used in testing and local development only.

**Comment by SpencerMarks-IDHA** [comment]
It would be good not have to override the static analysis tools though.

From: Jim ***@***.***>
Date: Monday, October 16, 2023 at 2:27 PM
To: globaldothealth/federation ***@***.***>
Cc: Marks, Spencer ***@***.***>, Review requested ***@***.***>
Subject: Re: [globaldothealth/federation] Replacing custom encryption with SSL (PR #11) [EXTERNAL]
* External Email - Caution *


Just don't print the certs and we are good here.

What remains is diagnostics for mocks used in testing and local development only.

**Comment by jim-sheldon** [comment]
> It would be good not have to override the static analysis tools though.

I will look into how to disable them for certain files to avoid false positives. I changed those log statements for the time being.

**Review by SpencerMarks-IDHA — APPROVED** [review]

**Pull request merged**